File: realtime/src/mqtt-bridge.ts
/**
 * MQTT → Socket.io bridge.
 * Subscribes to gdf/# on EMQX and emits sensor:reading events to
 * the per-sensor Socket.io room so only subscribed clients receive them.
 */
import mqtt, { MqttClient } from "mqtt";
import { Server } from "socket.io";
import { SensorReading } from "./types";

const MQTT_HOST = process.env.MQTT_HOST ?? "localhost";
const MQTT_PORT = parseInt(process.env.MQTT_PORT ?? "1883");
const MQTT_USERNAME = process.env.MQTT_USERNAME ?? "gdf_realtime";
const MQTT_PASSWORD = process.env.MQTT_PASSWORD ?? "realtime_secret";
const TOPIC_PREFIX = "gdf";

export function startMqttBridge(io: Server): MqttClient {
  const client = mqtt.connect(`mqtt://${MQTT_HOST}:${MQTT_PORT}`, {
    username: MQTT_USERNAME,
    password: MQTT_PASSWORD,
    clientId: `gdf-realtime-gateway-${Math.random().toString(16).slice(2, 8)}`,
    reconnectPeriod: 3000,
    keepalive: 60,
  });

  client.on("connect", () => {
    console.log(`[MQTT] Connected to ${MQTT_HOST}:${MQTT_PORT}`);
    client.subscribe(`${TOPIC_PREFIX}/#`, { qos: 1 }, (err) => {
      if (err) console.error("[MQTT] Subscribe error:", err);
      else console.log(`[MQTT] Subscribed to ${TOPIC_PREFIX}/#`);
    });
  });

  client.on("message", (topic: string, payload: Buffer) => {
    // Topic format: gdf/{sensor_id}/{channel}
    const parts = topic.split("/");
    if (parts.length !== 3 || parts[0] !== TOPIC_PREFIX) return;

    const [, sensor_id, channel] = parts;

    let value: number;
    let unit = "";
    let ts = new Date().toISOString();

    try {
      const data = JSON.parse(payload.toString());
      if (typeof data === "number") {
        value = data;
      } else {
        value = parseFloat(data.value);
        unit = data.unit ?? "";
        ts = data.ts ?? ts;
      }
    } catch {
      value = parseFloat(payload.toString());
      if (isNaN(value)) return;
    }

    const reading: SensorReading = { sensor_id, channel, value, unit, ts };

    // Emit to the sensor-specific room so only subscribed clients receive it
    io.to(`sensor:${sensor_id}`).emit("sensor:reading", reading);

    // Also emit to the "all" room for the main dashboard grid
    io.to("all").emit("sensor:reading", reading);
  });

  client.on("error", (err) => console.error("[MQTT] Error:", err));
  client.on("reconnect", () => console.log("[MQTT] Reconnecting..."));
  client.on("offline", () => console.warn("[MQTT] Client offline"));

  return client;
}

File: backend/app/core/mqtt_bridge.py
"""
MQTT bridge: subscribes to gdf/# on EMQX, decodes messages,
writes to TimescaleDB, and publishes to Redis for live dashboard.
"""
import asyncio
import json
import logging
from datetime import datetime, timezone

import aiomqtt

from app.core.config import settings
from app.core.redis_client import publish_reading

logger = logging.getLogger(__name__)


async def _handle_message(topic: str, payload: bytes, db_session_factory):
    """Parse gdf/{sensor_id}/{channel} messages and persist them."""
    parts = topic.split("/")
    if len(parts) != 3 or parts[0] != "gdf":
        return

    _, sensor_id, channel = parts

    try:
        data = json.loads(payload)
        value = float(data.get("value", data))  # support {"value":x} or bare number
        unit = data.get("unit", "")
        ts = data.get("ts", datetime.now(timezone.utc).isoformat())
    except (ValueError, TypeError):
        logger.warning("Unparseable MQTT payload on %s: %r", topic, payload)
        return

    # Persist to TimescaleDB
    async with db_session_factory() as session:
        from app.services.ingestion import insert_reading
        await insert_reading(session, sensor_id, channel, value, unit, ts)

    # Publish to Redis for Socket.io live push
    await publish_reading(sensor_id, channel, value, unit, ts)


async def run_mqtt_bridge(db_session_factory):
    """Long-running task: connects to EMQX and processes all sensor messages."""
    while True:
        try:
            async with aiomqtt.Client(
                hostname=settings.mqtt_host,
                port=settings.mqtt_port,
                username=settings.mqtt_username,
                password=settings.mqtt_password,
                identifier="gdf-backend-bridge",
            ) as client:
                logger.info("MQTT bridge connected to %s:%d", settings.mqtt_host, settings.mqtt_port)
                await client.subscribe(f"{settings.mqtt_topic_prefix}/#", qos=1)
                async for message in client.messages:
                    topic = str(message.topic)
                    asyncio.create_task(_handle_message(topic, message.payload, db_session_factory))
        except aiomqtt.MqttError as e:
            logger.error("MQTT connection error: %s — retrying in 5s", e)
            await asyncio.sleep(5)
